### PR TITLE
Initial proposal for git force pushing.

### PR DESCRIPTION
**Pull request by lededje**

### CODE REVIEW

File: _guidelines/git.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+When force pushing **always** use `--force-with-lease` to ensure there are no remote changes you are will destroy by accident.
+
+Remember, when performing _destructive_ commands that effect remote: **Communication is key.**
+
+- [Force Pushing Master:](#force-pushing--master) **Nope.** This is something that should never be done by anyone. Master should always be protected from force pushing and this should **never** be removed. If you need to undo something that is in master **revert** commits that are causing you issues and pr them in.
+- [Force Pushing Your Own Branch:](#force-pushing--own-branch) **Sure.** If no one else is working on it and you created the branch then force push at will. If someone has come and asked you if they can add something to your branch then check with them *before* force pushing anything.
+- [Force Pushing Someone Else's Branch](#force-pushing--own-branch) **Sometimes.** Talk to them about this before modifying their branch to ensure you know the full situation before force pushing.
```
Comment (gregbeech) [inline _guidelines/git.md]: This label is the same as the one above,

**Comment by martyphee** [comment]
:shipit:

**Comment by tompave-roo** [comment]
Makes sense.
There might be cases where you _must_ force-push to `master`, but they are such extraordinary situations that we don't need to specify any details here.

As per Greg's comment, one of the `#anchors` is duplicated.

Other than that, LGTM

**Comment by lededje** [comment]
+1 @tompave-roo In a _must_ force-push master case (thinking we accidentally push a private key* or something) then I would imagine a specific rollout for that case would be drawn up and executed.

*Obviously anything pushed should be considered compromised, but in an instance for example of pushing a private key that would take a short while to propagate would still be a good shout.

File: _guidelines/git.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+## Force Pushing
+
+Force pushing is one of the **destructive** functions of git in that it unconditionally overwrites the remote repository with whatever you have locally, possibly overwriting any changes that a team member has pushed in the meantime.
+
+When force pushing **always** use `--force-with-lease` to ensure there are no remote changes you are will destroy by accident.
+
+Remember, when performing _destructive_ commands that effect remote: **Communication is key.**
+
+- [Force Pushing Master:](#force-pushing--master) **Nope.** This is something that should never be done by anyone. Master should always be protected from force pushing and this should **never** be removed. If you need to undo something that is in master **revert** commits that are causing you issues and pr them in.
+- [Force Pushing Your Own Branch:](#force-pushing--own-branch) **Sure.** If no one else is working on it and you created the branch then force push at will. If someone has come and asked you if they can add something to your branch then check with them *before* force pushing anything.
```
Comment (tommyp) [inline _guidelines/git.md]: It's worth mentioning here that you shouldn't force pushing to your own branch if it's been merged to staging or if someone else has branched off from it.

Comment (tompave-roo) [inline _guidelines/git.md]: Good point.
Considering our current QA strategy, I would also stress the importance of staging.

**Comment by kapowaz** [comment]
One thing that’s missing from the sections concerning force pushing to branches other than master is something akin to a _service level agreement_: if another developer takes over the bulk of the work, or if the branch remains inactive for a long time, then at what point is it reasonable to assume a PR is abandoned, or that you can safely assume control of it (for merging purposes, say)?

I also think it’s slightly misleading to couch branches in terms of _ownership_; whilst a branch might have an initial author, the more collaborative a PR is (exactly the scenario in which rebasing might take place) the less true it is that there is any single canonical ‘owner’.

Ultimately the point about _communication is key_ is the most important thing for me in these guidelines — people need to understand that force pushing is destructive, and ensure that they understand exactly what they’re writing over by talking to other authors in the branch.

**Comment by tommyp** [comment]
+1 to @kapowaz's comment. Definitely hit home that communication with other Devs interested in your branch is key. Further to this, will this page cover pull requests or is that going to be another section?

**Comment by kapowaz** [comment]
@tommyp I imagine this page can be used for all good git-fu — you could always create another branch on top of this one to add your own section?

**Comment by lededje** [comment]
@kapowaz @tommyp RE: Zombie branches and taking over ownership:

The same rules should apply. The original author and any other contributors should be kept in the loop. If they are unreachable then fair enough.

**Comment by tommyp** [comment]
LGTM

File: _guidelines/git.md [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+- [Force Pushing Someone Else's Branch](#force-pushing--other-branch) **Sometimes.** Talk to them about this before modifying their branch to ensure you know the full situation before force pushing.
+- [Force Pushing a Merged Branch](#force-pushing--merged-branch) **Nope.** Force pushing a branch that has been merged to a branch that has a large exposure (for example staging or master) is not recommended. Merging to that branch again will be messy and problematic, and force pushing the branch with exposure will be even more so.
```
Comment (kapowaz) [inline _guidelines/git.md]: 👍

File: _guidelines/git.md [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+- [Force Pushing to Remove Sensitive Data](#force-pushing--sensitive-data) **Sure.** If sensitive data has been pushed by accident (private keys, stats or other company secrets) force push in accordance with the other rules (where possible) to expunge them from git history. Know that **any** information pushed should be deemed compromised and that anyone with a checkout of the code may retain a copy on their machine and any third parties (Github, Travis etc) whom might have caching that keeps the data alive. Notify senior staff of the issue and coordinate with them to get it sorted.
```
Comment (kapowaz) [inline _guidelines/git.md]: I wonder if this shouldn’t be its own section, given how this document will eventually serve as a general-purpose git usage guidelines doc. Accidentally pushing sensitive info and what to do to fix it is a scenario that could potentially be expanded out (a bit like how [Oh shit, git!](http://ohshitgit.com) has scenarios / git usage to resolve an issue), maybe?

Comment (lededje) [inline _guidelines/git.md]: I think this is a good start. Feel free to pr this to improve on it.

**Pull request merged**